source/logic.rst

```diff
@@ -0,0 +1,492 @@
+on the values of ``f``.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+    -- END
+
+In the next example, remember that ``λ x, f x + g x`` is the
+function that maps ``x`` to ``f x + g x``.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+
+    variables (f g : ℝ → ℝ) (a b : ℝ)
+
+    -- BEGIN
+    example (hfa : fn_ub f a) (hgb : fn_ub g b) :
+      fn_ub (λ x, f x + g x) (a + b) :=
+    begin
+      intro x,
+      dsimp,
+      apply add_le_add,
+      apply hfa,
+      apply hgb
+    end
+    -- END
+
+Applying ``intro`` to the predicate ``fn_ub (λ x, f x + g x) (a + b)``
+forces Lean to unfold the definition of ``fn_ub``
+and introduce ``x`` for the universal quantifier.
+The goal is then ``(λ (x : ℝ), f x + g x) x ≤ a + b``.
+But applying ``(λ x, f x + g x)`` to ``x`` should result in ``f x + g x``,
+and the ``dsimp`` command performs that simplification.
+(The "d" stands for "definitional.")
+You can delete that command and the proof still works;
+Lean would have to perform that contraction anyhow to make
+sense of the next ``apply``.
+The ``dsimp`` command simply makes the goal more readable
+and helps us figure out what to do next.
+The rest of the proof is routine.
+The last two ``apply`` commands force Lean to unfold the definitions
+of ``fn_ub`` in the hypotheses.
+Try carrying out similar proofs of these:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+
+    variables (f g : ℝ → ℝ) (a b : ℝ)
+
+    -- BEGIN
+    example (hfa : fn_lb f a) (hgb : fn_lb g b) :
+      fn_lb (λ x, f x + g x) (a + b) :=
+    sorry
+
+    example (nnf : fn_lb f 0) (nng : fn_lb g 0) :
+      fn_lb (λ x, f x * g x) 0 :=
+    sorry
+
+    example (hfa : fn_ub f a) (hfb : fn_ub g b)
+        (nng : fn_lb g 0) (nna : 0 ≤ a) :
+      fn_ub (λ x, f x * g x) (a * b) :=
+    sorry
+    -- END
+
+Even though we have defined ``fn_ub`` and ``fn_lb`` for functions
+from the reals to the reals,
+you should recognize that the definitions and proofs are much
+more general.
+The definitions make sense for functions between any two types
+for which there is a notion of order on the codomain.
+Checking the type of the theorem ``add_le_add`` shows that it holds
+of any structure that is an "ordered additive commutative monoid";
+the details of what that means don't matter now,
+but it is worth knowing that the natural numbers, integers, rationals,
+and real numbers are all instances.
+So if we prove the theorem ``fn_ub_add`` at that level of generality,
+it will apply in all these instances.
+
+.. code-block:: lean
+
+    variables {α : Type*} {R : Type*} [ordered_cancel_add_comm_monoid R]
```

Comment:
Maybe remind our readers they already met square brackets in Section 2.2, although the actual explanation is still to come.